hips/hip-0019.md

```diff
@@ -0,0 +1,186 @@
+* Migrate the container images the Helm Chart relies on to the isolated registry (same for `dependencies`).
+* Update the Helm Chart with the new references (same for `dependencies`).
+* Push the modified Helm Chart to the isolated registry (same for `dependencies`).
+* Install the Helm Chart.
+
+If Helm provides support for an container images lock file, many of these issues could be just addressed as part of the standard Helm tooling, automating the process and simplifying the work needed to move applications between registries. 
+
+## Rationale
+
+During the drafting of how to prepare an accurate images lock file, several alternatives found on the Internet were considered and discarded.
+
+The first option one could think of is rendering the Helm Chart and parsing the images from the equivalent Kubernetes manifests. From that point, it would be relatively easy to do the rest. However, post-render attempts of getting the images list aren't be reliable, as the manifests change depending on the attributes of the values files, or the user input.
+
+Some vendors like ArtifactHub use [annotations](https://github.com/artifacthub/hub/blob/master/charts/artifact-hub/Chart.yaml#L80), others like OpenShift use their own [proprietary fields](https://github.com/mongodb/helm-charts/blob/39df2aa9bc77e9f34c02c18ab7dbac172e5134a1/charts/enterprise-operator/values-openshift.yaml) in values files. Both end up falling into the same basket: promoting self-made conventions that aren't either backed up by an official specification, or supported by Helm itself natively.
+
+A third approach would be trying to `grep` the images references out of the Chart files. Again, this ends up being an attempt of promoting self-made conventions in order to make the _grepping_ work. Discussing this poinrt, we brought Operators to the table. These components may rely on images which aren't declared on the helm files, but pulled directly from the source code.
+
+Thus, we came to the conclussion that:
+
+* [HIP-0015](https://github.com/helm/community/blob/main/hips/hip-0015.md) already pushes for declaring lists of container images through official `annotations`.
+* Annotations are user-friendly (easy to write) but not machine-friendly (lacks of specification, misses digests...).
+* An `images-lock.yaml` (or similarly named) would be backed up by a spec. plus it could contain other useful information such as the images' digests that can be used by supply chains to enforce additional good security practices.
+* Helm could internally generate this file out of the `annotations` proposed in `HIP-0015`.
+* Helm could apply an existing `images-lock.yaml` file to the _post-rendered_ manifests to enforce a BOM (just like others such as `npm` do with their `packages-lock.yml`).
+* By parsing this file, Helm could help users to migrate Charts between registries, either by providing new commands or through new features in the existing commands.
```

Comment:
This could also be developed as a plugin initially.

Reply:
Yes, this is a good point @joejulian . It might be worth actually to start this as a plugin.

Reply:
Took a while due to other priorities but we ended up following your advice @joejulian https://github.com/vmware-labs/distribution-tooling-for-helm